apps/src/sites/studio/pages/components/editorRedux.js

```diff
@@ -0,0 +1,127 @@
+/* global levelKeyList */
+
+import _ from 'lodash';
+
+function updatePositions(node) {
+  for (var i = 0; i < node.length; i++) {
+    node[i].position = i + 1;
+  }
+}
+
+let newStageId = -1;
+let newLevelId = -1;
+
+export default function reducer(state, action) {
+  let newState = _.cloneDeep(state);
```

Comment:
I would go one step further and suggest using `immutable-js` data structures for your state. It makes it a lot easier to efficiently create new state objects with slightly different data.